### DISCUSSION TITLE
Berry: tasmota.get_switches() is inverted, why?

### DISCUSSION THREAD

**Discussion by jeroenst** [arendst/Tasmota#18224] [Q&A]
When I use tasmota.get_switches() in berry I see that when an input is 1 the berry value is false and when the input is 0 the berry value is 1.

I read that this is on purpose in the manual:
```
tasmota.get_switches | () -> list(bool)Returns as many values as switches are present. true means PRESSED and false means NOT_PRESSED. (Warning: this is the opposite of the internal representation where PRESSED=0)
```

My question is, why is this chosen this way, it is confusing imo.

**Comment by sfromis** [comment]
What sounds confusing would be for `true` to mean *not pressed*, and `false` to mean *pressed*? I'd say that the confusing thing is that base Tasmota reports "ON" when a switch is not activated - and Berry improves on this :wink: 

While you of course *can* have inverted switches, the common setup is that you activate the input by connecting to ground.

**Reply by jeroenst** [reply]
At the Dingtian 8CH, when the switch is pressed when not pressed and false when it's pressed, so indeed it's very confusing.

This is when no switches are pressed:
<img width="436" height="242" alt="image" src="https://github.com/user-attachments/assets/56d12ddd-df61-460a-8c50-d90fd060ffc4" />

And this is when switch one is pressed:
<img width="436" height="242" alt="image" src="https://github.com/user-attachments/assets/43232b95-f57d-4963-89a7-55a4cc8c29da" />

**Reply by sfromis** [reply]
I suppose that the one coding the Digtian driver decided to have them match the behavior of inputs connected directly to a gpio. In this context, what you see is just normal and unsurprising :stuck_out_tongue: 

To me it would make sense to have them follow the gpio multiplexer drivers having a "version 2" setup without specialized payloads, to "simply" have the behavior of standard I/O gpios. This would "fix" the confusion by not cluttering the page with that "DINGTIAN" line.

**Reply by jeroenst** [reply]
I think a setoption (or a define) to invert the multiplexer inputs would be the most universal way.

**Reply by sfromis** [reply]
Just don't have a starting assumption about how you'd think of it as "making sense" with inherent meanings in seeing high/low/1/0/true/false/on/off around single-bit input states, but adapt to what you get. As basic switch/button input behavior (with the normal wiring connecting to ground when engaged, and pullup when not), input being high/1 primarily means not engaged. Very typical behavior around microcontrollers.

Any optional post-invert for your preferences would logically be for all such inputs, not only multiplexers (etc).

It works in a similar way with things like I2C or serial pins being high when inactive, and pulled low with signal. One explanation of this is that inputs can be "open drain", and no need for deciding when a signal should be 3.3V vs 5V. Open drain means that the receiver pulls up (according to its voltage level) until the signal pulls down to ground. Simplifies circuit design.

**Comment by jeroenst** [comment]
I would expect input HIGH = PRESSED = ON = TRUE = 1,
And when inverted input LOW = PRESSED = ON = TRUE = 1...

So I understand the design of tasmota where PRESSED = ON = FALSE = 0 is causing this?

**Reply by sfromis** [reply]
Don't know where you got your expectations mixed so much up. And you cannot "invert" a Switch pin, only use an inverted SwitchMode. Result of whatever SwitchMode you use is outside of this base consideration:

For a "normal" Switch input:

- Signal low (0V, pressed = connected to ground) is OFF, has no 0/1 value for a Switch. Becomes true for "pressed" in Berry.
- Signal high (~3.3V, released = unconnected but internally pulled up) is ON, not "0 or 1". Becomes false for "not pressed" in Berry.

If your signal is electrically inverted, things gets flipped, but that's c'est la vie. Tasmota cannot "see" the physical arrangement.

**Reply by jeroenst** [reply]
This is the case in pull up and the switch connected to ground. When an input is pulled down switch is connected to 3v3 then pressed (on,true,1) is 3v3.

The most confusing thing for me is that on the main page I see "DINGTIAN Input0..Input7	00000000" while in berry tasmota.get_switches() gives [true, true, true, true, true, true, true, true].

**Reply by sfromis** [reply]
As you are not talking about regular Switch inputs, but that Dingtian thing, that adds and extra level to the inherent confusion about what things mean around "switches".

As Tasmota does not provide for "inverted Switch", you'll basically have to ignore any expectations of what "0", "1", "ON", "OFF", "true" and "false" should mean in relation to available physical controls you are operating, but just observe. :eyeglasses:

**Comment by Wintespe** [comment]
From the documentation:
(https://tasmota.github.io/docs/Commands/#how-to-use-commands)
### How to use commands

Every command used without a parameter (payload) returns the current setting.

Power returns the status of first defined power output (usually Relay1)

Instead of 0 you can use off or false and instead of 1 you can use on or true.

Power ON turns first defined power output (usually Relay1) on

Power1 1 also turns first defined power output (usually Relay1) on

Power1 True also turns first defined power output (usually Relay1) on
